tools/toolchains/arm.py

```diff
@@ -91,6 +94,32 @@ def __init__(self, target, notify=None, macros=None,
 
         self.SHEBANG += " --cpu=%s" % cpu
 
+    def version_check(self):
+        stdout, _, retcode = run_cmd([self.cc[0], "--vsn"], redirect=True)
+        msg = None
+        min_ver, max_ver = self.ARMCC_RANGE
+        first_line = stdout.splitlines()[0]
+        match = self.ARMCC_VERSION_RE.match(first_line)
```

Comment:
There's no end anchor in the pattern. the difference between this and re.search becomes if it finds the match starting at the beginning of the string. I have "Prooduct: ARM Compiler 5.06u3 Professional" on my machine and the current code spits out 5.06 as the version.

Reply:
we are parsing a toolchain output. It is filled in by humans and we have no confidence that this string will remain unchanged between versions. Thus we will probably want less strict regex as it may/will change.

According to [Python documentation](https://docs.python.org/3/library/re.html#search-vs-match) (also true for Python 2.x) the difference between `re.match()` and `re.search()` is:

> `re.match()` checks for a match only at the beginning of the string, while `re.search()` checks for a match anywhere in the string

please note that you have `^` in the compiled regex (line 44). If tomorrow some very gifted developer will add change `Product` to `Best toolchain ever Product:` you will have to modify your code in two places.
I bet you will miss the `re.match` thing at a first attempt - removing `^` (the obvious change) won't help.
You (or someone less experienced with regex) will have to debug to understand why regex is not working.
My gut feeling is to use search - but it is your call.
And if you are searching - you do not have to `stdout.splitlines()` - just search for the pattern you expect and extract it. In this changeset you are using line 0, but for other files you do iterate over `splitlines()`